apps/src/code-studio/components/progress/LessonProgress.jsx

```diff
@@ -175,13 +185,44 @@ class LessonProgress extends Component {
                 isCurrent = currentPageNumber === level.pageNumber;
               }
 
-              // When the user clicks on a level bubble for which we support changing to
-              // that level without reloading the page, we use a click handler which
-              // calls through to the progress redux store to change to that level
-              // immediately.
-              const onBubbleClick = canChangeLevelInPage(currentLevel, level)
-                ? () => navigateToLevelId(level.id)
-                : undefined;
+              const isGenAiLevel = () => {
+                const path = new URL(document.location).pathname;
+                const pathComponents = path.split('/');
+                return pathComponents.includes('gen-ai-customizing-pilot-v1');
+              };
+
+              const hasUnsavedChanges = () => {
+                let hasUnsavedChanges = false;
+                if (
+                  this.props.savedAiCustomizations &&
+                  this.props.currentAiCustomizations
+                ) {
+                  hasUnsavedChanges =
+                    findChangedProperties(
+                      this.props.savedAiCustomizations,
+                      this.props.currentAiCustomizations
+                    ).length > 0;
+                }
+                return hasUnsavedChanges;
+              };
+
+              /**
+               * When the user clicks on a level bubble for which we support changing to
+               * that level without reloading the page, we use a click handler which
+               * calls through to the progress redux store to change to that level
+               * immediately. Unless we're on a GenAi level with unsaved changes, in which
+               * case we show a reminder to explicitly save the changes.
+               **/
+              const getOnBubbleClick = (level, currentLevel) => {
+                if (isGenAiLevel() && hasUnsavedChanges()) {
+                  console.log('show save reminder should be the function');
+                  return setShowSaveChangesReminder(true);
```

Comment:
@sanchitmalhotra126 This is my messy WIP. Could use a 🦆 please. 

a.) I'm still not sure this is the right implementation and 
b.) if it is the way to accomplish what we want, these components handle dispatch differently - do you have advice for how to work around that?

Reply:
Yeah this is super tricky for all the reasons you mentioned 🙂 there are so many paths to change levels and the signal that lab2 uses to listen for level change is just that the current level ID has updated. So doing anything that 1) knows when a level is changing and 2) also _prevents_ the level change from happening is difficult until and unless we have some singular central level switching system. And yep agreed that it's not ideal to have Gen AI (or any lab) - specific logic in the level progress layer.

Let's discuss at the Gen AI sync, but I think in the short term, it doesn't seem super feasible for us to block level changing in lab2. Longer term, we potentially could have a central level switching controller that all level change requests are routed through, and individual labs/other components could register handlers to respond to and optionally block level changes. But for now if it's alright for us to just show a banner/warning to users, that would be much more straightforward to implement.